server/auth_callout.go

```diff
@@ -0,0 +1,329 @@
+				respCh <- false
+				return
+			} else if !acc.isAllowedAcount(juc.Issuer) {
+				s.Warnf("Account %q not allowed for auth callout on %q", juc.Issuer, issuer)
+				respCh <- false
+				return
+			}
+		}
+		allowNow, validFor := validateTimes(juc)
+		if !allowNow {
+			c.Errorf("Outside connect times")
+			respCh <- false
+			return
+		}
+		allowedConnTypes, err := convertAllowedConnectionTypes(juc.AllowedConnectionTypes)
+		if err != nil && len(allowedConnTypes) == 0 {
+			respCh <- false
+			return
+		}
+		// Apply to this client.
+		targetAcc := acc
+		// Check if we are being asked to switch accounts.
+		if aname := juc.Audience; aname != _EMPTY_ {
+			targetAcc, err = s.LookupAccount(aname)
+			if err != nil {
+				s.Warnf("No valid account %q for auth callout response: %v", aname, err)
+				respCh <- false
+				return
+			}
+			// In operator mode make sure this account matches the issuer.
+			if isOperatorMode && aname != juc.Issuer {
+				s.Warnf("Account %q does not match issuer %q", aname, juc.Issuer)
+				respCh <- false
+				return
+			}
+		}
+
+		// Build internal user and bind to the targeted account.
+		nkuser := buildInternalNkeyUser(juc, allowedConnTypes, targetAcc)
+		if err := c.RegisterNkeyUser(nkuser); err != nil {
+			s.Warnf("Could not register auth callout user: %v", err)
+			respCh <- false
+			return
+		}
+
+		// See if the response wants to override the userID.
+		if juc.Name != _EMPTY_ {
+			c.mu.Lock()
+			c.opts.Username = juc.Name
+			// Clear any others.
+			c.opts.Nkey = _EMPTY_
+			c.pubKey = _EMPTY_
+			c.opts.Token = _EMPTY_
+			c.mu.Unlock()
+		}
+
+		// Check if we need to set an auth timer if the user jwt expires.
+		c.setExpiration(juc.Claims(), validFor)
+
+		respCh <- true
+	}
+
+	sub, err := acc.subscribeInternal(reply, processReply)
+	if err != nil {
+		s.Warnf("Error setting up reply subscription for auth request: %v", err)
+		return false
+	}
+	defer acc.unsubscribeInternal(sub)
+
+	// Build our request claims.
+	claim := jwt.NewAuthorizationRequestClaims(AuthRequestSubject)
+	// Set expected public user nkey.
+	claim.UserNkey = pub
+
+	s.mu.RLock()
+	claim.Server = jwt.ServerID{
+		Name: s.info.Name,
+		Host: s.info.Host,
+		ID:   s.info.ID,
+	}
+	// Check if we have been requested to encrypt.
+	if xkp != nil {
+		claim.Server.XKey = xkey
+	}
+	s.mu.RUnlock()
+
+	authTimeout := secondsToDuration(s.getOpts().AuthTimeout)
+	claim.Expires = time.Now().Add(time.Duration(authTimeout)).UTC().Unix()
+	if opts.AuthCallout != nil {
+		claim.Audience = opts.AuthCallout.Issuer
+	} else {
+		claim.Audience = acc.Name
+	}
+
+	// Grab client info for the request.
+	c.mu.Lock()
+	c.fillClientInfo(&claim.ClientInformation)
+	c.fillConnectOpts(&claim.ConnectOptions)
+	// If we have a sig in the client opts, fill in nonce.
+	if claim.ConnectOptions.SignedNonce != _EMPTY_ {
+		claim.ClientInformation.Nonce = string(c.nonce)
+	}
+
+	// TLS
+	if c.flags.isSet(handshakeComplete) && c.nc != nil {
```

Comment:
Not necessarily for the first pass, but wondering if this should be driven by account configuration.  At scale this could result in a fair amount of extra data an auth callout wouldn't necessarily use.

Reply:
Let's wait and see, not that much really compared to production systems that I have seen send 100k msgs/sec that are 4-8k.